### PR TITLE
editing through

### PR DESCRIPTION
**Pull request by petrelharp**
opening now for future discussion

### CODE REVIEW

File: forwards_paper.tex [hunk 1]
```diff
@@ -63,18 +63,20 @@
...
-    can be stored on $O(N \log N + M)$ space, making it feasible to store a simulation's entire final generation
+    can be stored in $O(N \log N + M)$ space, making it feasible to store a simulation's entire final generation (and its' history)
     rather than only a subset.
-    (3) A simulation can easily be initialized with prior history produced by a more efficient coalescent simulation of deep history.
+    (3) A simulation can easily be initialized with a more efficient coalescent simulation of deep history.
     To make these possible,
-    we introduce the \msprime{} Tables API, which allows for the efficient interchange of genealogical data between modular program components.
+    we introduce the \msprime{} Tables API, which allows for the efficient interchange 
+    of tree sequences between modular program components,
+    and show how to quickly simplify a tree sequence by removing history irrelevant to a given set of genomes.
```
Comment (petrelharp) [inline forwards_paper.tex:80]: thought we should mention 'simplify' here

File: forwards_paper.tex [hunk 2]
```diff
@@ -128,57 +131,57 @@ \section*{Introduction}
 -- and often, the point of such simulations is to compare to real data --
 then such simulations must record each individual's genome.
 As samples of most species' genomes harbor tens or hundreds of millions of variant sites,
-naively carrying full genotypes for even modest numbers of individuals through a simulation
-becomes quickly prohibitive.
+carrying full genotypes for even modest numbers of individuals through a simulation
+can quickly become prohibitive.
```
Comment (petrelharp) [inline forwards_paper.tex:135]: I believe this statement is true even if not talking about "naively"?

File: forwards_paper.tex [hunk 3]
```diff
@@ -147,38 +151,37 @@ \section*{Introduction}
...
 However, it is thought that most genetic variation is selectively neutral (or nearly so).
-By definition, the neutral alleles carried by individuals in a population
+By definition, neutral alleles carried by individuals in a population
 do not affect the population process.
 For this reason, if one records the entire genealogical history of a population over the course of a simulation,
 simply laying down neutral mutations on top of that history afterwards
-is equivalent to having generated them during the simulation.
-Precisely, we would need to know the genealogical tree relating all sampled individuals
+is equivalent to having generated them during the simulation:
+it does not matter if we generate each generation's mutations during the simulation, or afterwards.
```
Comment (petrelharp) [inline forwards_paper.tex:159]: trying to explain this point about "can just throw them down afterwards" more better as it's not obvious to everyone

File: forwards_paper.tex [hunk 4]
```diff
@@ -147,38 +151,37 @@ \section*{Introduction}
...
+Precisely, we need to know the genealogical trees relating all sampled individuals
 at each position along the genome.
-This is equivalent to knowing the \emph{population pedigree} --
-the complete history of parent-offspring relationships of an entire population
-going back to a remote time -- and the genetic outcomes of each ancestral
-meiosis.
-The embellished pedigree contains all the information necessary
-to construct the genealogical tree that relates each individual to every other
-at each position on the genome,
-i.e., the \emph{tree sequence} implemented in the coalescent simulator \msprime{}~\citep{kelleher2016efficient}.
```
Comment (petrelharp) [inline forwards_paper.tex:164]: I rearranged this bit to (hopefully) flow more logically, separating "the embellished population pedigree" from "the tree sequence".

File: forwards_paper.tex [hunk 5]
```diff
@@ -147,38 +151,37 @@ \section*{Introduction}
...
-simulation.
+The information in this embellished pedigree is stored as a \emph{succinct tree sequence} 
+(or, for brevity, ``tree sequence''),
+which contains all the information necessary
+to construct the genealogical tree that relates each individual to every other
+at each position on the genome.
+
+In this paper, we describe a storage method for \emph{succinct tree sequences} 
+and, hence, genome sequence, as well as an algorithm for simplifying these.
+The data structure is \emph{succinct} in the sense that its' space usage is close to optimal,
```
Comment (petrelharp) [inline forwards_paper.tex:177]: succinct here

File: forwards_paper.tex [hunk 6]
```diff
@@ -187,35 +190,40 @@ \section*{Introduction}
 \section*{Results}
 
 
-We begin by benchmarking the performance improvement achieved by connecting
-a previously published forwards-time simulation library,
-\fwdpp{} \citep{fwdpp},
-to the tools we describe here, as implemented in \msprime{}.
+The strategy described above is only of interest if it is computationally feasible.
```
Comment (petrelharp) [inline forwards_paper.tex:193]: trying to say *why* we put the simulation results before saying how they actually work

File: forwards_paper.tex [hunk 7]
```diff
@@ -187,35 +190,40 @@ \section*{Introduction}
...
+retrospectively, as described below, resulting in statistically identical simulation results.
 % Both mutations and recombinations occur as Poisson processes with respective means $\mu$ and $r$,
 % and we used the infinitely-many sites mutation model \citep{Kimura1969-uc}.
-We ran simulations with different values of $N$ and varied the size of the region according to the scaled recombination
-parameter $\rho = 4Nr$.  When simulating neutral mutations, we kept the scaled mutation parameter, $\theta = 4N\mu$
-equal to $\rho$.
+We ran simulations with different values of $N$ and varied the size of the genomic region according to the scaled recombination
+parameter $\rho = 4Nr$.  
+%% this was said already:
+% When simulating neutral mutations, we kept the scaled mutation parameter, $\theta = 4N\mu$, equal to $\rho$.
```
Comment (petrelharp) [inline forwards_paper.tex:226]: since $\mu = \rho$ this follows; and we don't use $\theta$ elsewhere.

File: forwards_paper.tex [hunk 8]
```diff
@@ -225,40 +233,23 @@ \subsection*{Simulation benchmarks}
 coalescent \citep{Neuhauser1997-nn}.  Therefore, these selected mutations must be explicitly tracked in our forward simulation
 and the time savings due to pedigree recording come from not having to record \textit{neutral} mutations.
 
-We implemented our method in \cpp{}, using \fwdpp{} library functions and
```
Comment (petrelharp) [inline forwards_paper.tex:228]: moved this bit to Methods.

File: forwards_paper.tex [hunk 9]
```diff
@@ -225,40 +233,23 @@ \subsection*{Simulation benchmarks}
...
-hyperthreading enabled.
-We ran one simulation at a time and the machine was under minimal load otherwise.
-We used GNU parallel \citep{Tange2011a} to kill any simulation that did not finish within 72 hours,
-and the Linux \texttt{time} command to record run time and peak memory usage of each replicate.
-%% Our simulation scripts separately recorded the time spent in various steps of each run.
-
-%% \krt{currently, we are not doing anything with the detailed timings.  I think that is fine, and that we can delete last
-%%sentence of preceding paragraph.}
-
-
```
Comment (petrelharp) [inline forwards_paper.tex:244]: just ironed out these next paragraphs

File: forwards_paper.tex [hunk 10]
```diff
@@ -225,40 +233,23 @@ \subsection*{Simulation benchmarks}
...
-The total run times for simulations with selection are shown in Figure~\ref{fig:runtimes_selection}.  When tracking
-neutral mutations (instead of the pedigree), run times increase dramatically with increasing region size ($4Nr = 4Nu$ for
-these simulations).  Simulations with $N=5 \times 10^4$ timed out for region sizes larger than $10^3$.
```
Comment (petrelharp) [inline forwards_paper.tex:247]: moved this note about timing out to the figure caption so it doesn't get in the way of the flow

File: forwards_paper.tex [hunk 11]
```diff
@@ -293,45 +295,31 @@ \subsection*{Simulation benchmarks}
 
 
 %%%%%%
-\subsection*{Working with a tree sequence through tables}
+\subsection*{Tables for succinct tree sequences}
```
Comment (petrelharp) [inline forwards_paper.tex:298]: this title ("Working with") wasn't speaking to me; besides it is more about *storing*.

File: forwards_paper.tex [hunk 12]
```diff
@@ -293,45 +295,31 @@ \subsection*{Simulation benchmarks}
...
 
-The simulations above recorded information about each new individual
-in a collection of tables that together define a \emph{tree sequence}.
+We now explain what we actually did to achieve a $50\times$ speedup.
+The ``pedigree recording'' simulations above recorded information about each new individual
+in a collection of tables that together define a \emph{succinct tree sequence}
+(or, simply ``tree sequence'').
 A {tree sequence} is an encoding for a sequence of correlated trees,
 such as those describing the history of a sexual population.
-A closely related structure is known as the \emph{ancestral recombination graph} --
```
Comment (petrelharp) [inline forwards_paper.tex:302]: moved ARG connection to the Discussion

File: forwards_paper.tex [hunk 13]
```diff
@@ -293,45 +295,31 @@ \subsection*{Simulation benchmarks}
...
-ARGs are rarely used
-in practise. In contrast, the algorithmic properties of tree sequence
-algorithms have been explored in detail~\citep{kelleher2016efficient},
-contributing substantially to the efficiency of the \msprime\ coalescent simulator.
-
 Tree sequences are efficient because branches that are shared by adjacent trees are stored once,
 rather than repeatedly for each tree.
 The topology of a tree sequence is defined via its \emph{nodes} and \emph{edges},
-while information about variants are recorded as \emph{sites} and \emph{mutations}.
-The format for these four tables is given in more detail in the Methods;
```
Comment (petrelharp) [inline forwards_paper.tex:320]: there is no such detail in the Methods

File: forwards_paper.tex [hunk 14]
```diff
@@ -293,45 +295,31 @@ \subsection*{Simulation benchmarks}
...
-here we explain conceptually,
-using the example of Figure~\ref{fig:example_tree_sequence}. This formulation
-is derived from the ``coalescence records'' encoding of tree
+while information about variants is recorded as \emph{sites} and \emph{mutations};
+we give an example in Figure~\ref{fig:example_tree_sequence}. 
+This formulation is derived from the ``coalescence records'' encoding of tree
 sequences~\citep{kelleher2016efficient}, normalised to remove redundancy
-and generalised to include a much larger class of tree topology.
```
Comment (petrelharp) [inline forwards_paper.tex:325]: still countably infinite ;)

File: forwards_paper.tex [hunk 15]
```diff
@@ -293,45 +295,31 @@ \subsection*{Simulation benchmarks}
...
+and generalised to include a more general class of tree topologies.
 
 The \emph{nodes} of a tree sequence
-correspond to the vertices in the individual genealogies along the sequence,
-and each node refers to a distinct ancestor.
-Since each node represents a specific ancestor, it has a unique ``time'',
-thought of as her birth time, which determines the height of any vertices
```
Comment (petrelharp) [inline forwards_paper.tex:331]: nodes are haploid, so don't have a sex (and I think it reads better with 'the node')

File: forwards_paper.tex [hunk 16]
```diff
@@ -449,32 +439,33 @@ \subsection*{The \msprime\ Tables API}
...
 values in memory are from the same column, they tend to compress well,
 and suitable encodings can be chosen on a per-column basis~\citep{abadi2006integrating}.
 A particular advantage of this approach is that it enables very
 efficient copying of data, and in principle zero-copy data access
-(where a data consumer reads directly from the memory of a producer) is possible.
-Our implementation uses the NumPy C API~\citep{walt2011numpy} to efficiently copy
-data from Python into the low-level C library used to manipulate tree sequences.
+(where a data consumer reads directly from the memory of a producer).
+Our implementation 
+% uses the NumPy C API~\citep{walt2011numpy} to efficiently copy
```
Comment (petrelharp) [inline forwards_paper.tex:455]: modified re Kevin's comment

File: forwards_paper.tex [hunk 17]
```diff
@@ -484,50 +475,26 @@ \subsection*{The \msprime\ Tables API}
...
-% which is processed by the \msprime{} tables API.
-% Our implementation also uses \texttt{pybind11}
-% to create a Python package extending \texttt{fwdpy11}
-% (a Python package based on \fwdpp{}; \url{https://github.com/molpopgen/fwdpy11}),
-% % (A neat aside, that we may wish to skip here, is that it may be possible to use the tables
-% % API from a command-line C/\cpp{} application via an embedded Python interpreter.)
-
+Benchmarking results from this implementation are shown (alongside results from \fwdpp{})
+for simulations without selection in Appendix~\ref{ss:timing_nosel}:
+a relatively modest speedup of around $5 \times$ is achieved, likely due to increased overhead.
```
Comment (petrelharp) [inline forwards_paper.tex:487]: I think this ("likely") is fair to say?

File: forwards_paper.tex [hunk 18]
```diff
@@ -568,15 +535,15 @@ \subsection*{Recording the pedigree in forwards time}
 }
 
 \algstep{W4.}{Record edges.}{
-Call $\taddedge{\Et, 0, x, P_a, u}$ and $\taddedge{\Et, x, r, P_b, u}$.
+Call $\taddedge{\Et, 0, x, P_a, u}$ and $\taddedge{\Et, x, L, P_b, u}$.
```
Comment (petrelharp) [inline forwards_paper.tex:538]: whoopsie

File: forwards_paper.tex [hunk 19]
```diff
@@ -625,17 +593,14 @@ \subsection*{Recording the pedigree in forwards time}
 \end{figure}
 
 
-\krt{The following paragraph has a sentence mangling.  I think some comment is warrented about when it is useful to
-record selected mutations, like for ancient samples?}
```
Comment (petrelharp) [inline forwards_paper.tex:629]: I've got something, but not about ancient samples.

File: forwards_paper.tex [hunk 20]
```diff
@@ -776,8 +749,7 @@ \subsection*{Tree sequence simplification}
 (and so $u$ corresponds to a ``color'' in the description above).
 % These segments are stored in a collection of linked lists,
 % one list of segments for each ancestral lineage present at that time.
-We also maintain a map $A$ such that $A_j$ is the lineage
-corresponding to node $j$ in the input tree sequence.
+We also maintain a map from input nodes to lineages. % called $A_j$
```
Comment (petrelharp) [inline forwards_paper.tex:752]: we don't use this notation ($A_j$) elsewhere

File: forwards_paper.tex [hunk 21]
```diff
@@ -915,41 +890,38 @@ \subsubsection*{Estimates of run-time complexity}
...
-in our scheme we must store $13N$ quantities per generation (two edges and one node per individual).
-\plr{XXX}
-The simplification algorithm is linear in the number of edges of the input tree sequence
-(because it must process each edge),
-and so multiplies this by a constant factor.
-These considerations imply that propagating neutral genotypes for $T$ generations has time complexity $O(\mu T N^2)$,
-while our implementation is only $O(T N \log(2N))$.
+which grows with $\log(n)$, as shown in Figure~\ref{fig:simplify_complexity}C,
+
+
```
Comment (petrelharp) [inline forwards_paper.tex:909]: This paragraph below doesn't have much content: it compares to the naive method, which no-one good uses, and doesn't justify things.

File: forwards_paper.tex [hunk 22]
```diff
@@ -972,17 +944,13 @@ \section*{Discussion}
 It was implemented completely in AnA-FiTS~\citep{aberer2013rapid},
 but without the critical step of discarding irrelevant genealogical information.
 A similar but much more limited method for
-discarding this information does appear in \citet{padhukasahasram2008exploring}.
-
-
-\krt{The following para contains numbers that are results, not discussion.}
```
Comment (petrelharp) [inline forwards_paper.tex:978]: I think that's OK.

File: forwards_paper.tex [hunk 23]
```diff
@@ -1016,26 +984,44 @@ \section*{Discussion}
 The one-way nature of information flow here makes our scheme relatively easy to incorporate
 into a parallel algorithm.
 Recently, \citet{lawrie2017accelerating} used parallelization
-to improve performance for the ``Poisson Random Field'' (PRF) family of models~\citep{Sawyer1992-jw},
+to improve performance for the ``Poisson Random Field'' family of models~\citep{Sawyer1992-jw},
 in which mutations are unlinked and there is no effect of genetic background on fitness.
-However, thus far little attention has been paid to parallelization of simulations involving linkage, as our simulations
-where do, which is much more
-complex than the PRF model.
+However, thus far little attention has been paid to parallelization of more complex simulations.
```
Comment (petrelharp) [inline forwards_paper.tex:989]: The last sentence was hard to parse; hopefully this is a good substitute?

File: forwards_paper.tex [hunk 24]
```diff
@@ -1016,26 +984,44 @@ \section*{Discussion}
...
+theory~\citep{wiuf1997number,wiuf1999ancestry,marjoram2006coalescent,wilton2015smc}.
+However, the properties of the ARG as a computational structure have not
+been studied and, despite several efforts to standardise a common
+format~\citep{morin2006netgen,mcgill2013graphml}, % TODO check these refs against others in msprime paper
+ARGs are rarely used in practise. 
+In contrast, the algorithmic properties of tree sequence
+algorithms have been explored in detail~\citep{kelleher2016efficient},
+contributing substantially to the efficiency of the \msprime{} coalescent simulator.
+
+\paragraph{Bayesian phylogenetics}
```
Comment (petrelharp) [inline forwards_paper.tex:1006]: Short note.  Nice, but not much there.

File: forwards_paper.tex [hunk 25]
```diff
@@ -1054,24 +1040,28 @@ \section*{Discussion}
...
+pybind11 \citep[version 2.2.1][]{pybind11},
+and GCC (version 4.8.5).
+We ran all benchmarks on an Ubuntu Linux (version 16.04) system with two 2.6 GHz Intel E5-2650 CPU with
+hyperthreading enabled.
+We ran one simulation at a time and the machine was under minimal load otherwise.
+We used GNU parallel \citep{Tange2011a} to kill any simulation that did not finish within 72 hours,
+and the Linux \texttt{time} command to record run time and peak memory usage of each replicate.
+%% Our simulation scripts separately recorded the time spent in various steps of each run.
 
-\begin{enumerate}
```
Comment (petrelharp) [inline forwards_paper.tex:1061]: I got these in above

File: forwards_paper.tex [hunk 26]
```diff
@@ -1054,24 +1040,28 @@ \section*{Discussion}
...
-\end{enumerate}
 
 %%%%%%%%%%%%%%%%%%%%%%
 \section*{Acknowledgements}
-Thanks to Brad Shaffer and Evan McCartney--Melstad for useful discussions.
-Funding for this project came from XXX
+Thanks to Jared Galloway, Brad Shaffer, and Evan McCartney--Melstad for useful discussions.
+Work on this project was supported by funding from 
+the Sloan Foundation and the NSF (under DBI-1262645) to PR;
+and XXX.
```
Comment (petrelharp) [inline forwards_paper.tex:1064]: @jeromekelleher @ashander @molpopgen any funding you want to acknowledge?

File: sims/plot_RAM.py [hunk 27]
```diff
@@ -57,7 +57,7 @@
 
 ax_fwdpp.legend(loc='best',frameon=False)
 ax_fwdpp.set_title("With neutral mutations", fontsize='medium')
-ax_fwdpp_arg.set_title("With pedigree tracking", fontsize='medium')
```
Comment (petrelharp) [inline sims/plot_RAM.py:60]: s/tracking/recording/

**Comment by petrelharp** [comment]
I've edited through.  If I've undid some recent edit you made without noticing, apologies: let me know.

I left the "simplification" algorithm in the appendix alone, pending #59 .

**Comment by jeromekelleher** [comment]
Looks good to me. I'm going to merge this and follow up with my own edits.

File: forwards_paper.tex (hunk 5 above)
Comment (jeromekelleher) [reply forwards_paper.tex:177]: Perfect. I'll add a reference for succinct data strutures.

File: forwards_paper.tex (hunk 15 above)
Comment (jeromekelleher) [reply forwards_paper.tex:331]: agreed

File: forwards_paper.tex (hunk 16 above)
Comment (jeromekelleher) [reply forwards_paper.tex:455]: Msprime does use the numpy C api though. I'll update this to be clearer.

Comment (jeromekelleher) [reply forwards_paper.tex:455]: Actually what's there is good, ignore this comment.

File: forwards_paper.tex (hunk 17 above)
Comment (jeromekelleher) [reply forwards_paper.tex:487]: agreed.

File: forwards_paper.tex (hunk 18 above)
Comment (jeromekelleher) [reply forwards_paper.tex:538]: Good spot!

File: forwards_paper.tex (hunk 19 above)
Comment (jeromekelleher) [reply forwards_paper.tex:629]: Looks good.

File: forwards_paper.tex (hunk 26 above)
Comment (ashander) [reply forwards_paper.tex:1064]: USFWS but not sure of the contract number. Will email
